### PR TITLE
ESP32-S2: Add SPI Slave support

### PR DESCRIPTION
**Pull request by JurajSadel**
## Thank you for your contribution!

We appreciate the time and effort you've put into this pull request.
To help us review it efficiently, please ensure you've gone through the following checklist:

### Submission Checklist 📝
- [x] I have updated existing examples or added new ones (if applicable).
- [x] I have used `cargo xtask fmt-packages` command to ensure that all changed code is formatted correctly.
- [x] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
#### Extra:
- [x] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/CONTRIBUTING.md) and followed its instructions.

#### Description
Adds SPI Slave for ESP32S2 and removes ESP32 parts from it.

#### Testing
Tested with `spi_slave_dma` example.

Builds on top of #1514 but removed all ESP32-related changes from the driver. After discussion with @bjoernQ, we agreed it makes sense to add _just_ S2 and remove ESP32 parts, we've burnt so much time on that and the thing doesn't work.

### CODE REVIEW

File: esp-hal/src/spi/slave.rs [hunk 1]
```diff
@@ -423,32 +407,45 @@ where
         self.enable_dma();
 
         reset_dma_before_load_dma_dscr(reg_block);
-        tx.prepare_transfer_without_start(
```
Comment (MabezDev) [inline esp-hal/src/spi/slave.rs:426]: I don't know if this is the diff being weird, but did you mean to remove this block?

File: esp-hal/src/spi/slave.rs [hunk 2]
```diff
@@ -423,32 +407,45 @@ where
...
             .dma_conf()
             .modify(|_, w| w.dma_slv_seg_trans_en().clear_bit());
 
-        tx.start_transfer()?;
-        Ok(rx.start_transfer()?)
+        self.clear_dma_interrupts();
+        self.setup_for_flush();
+        reg_block.cmd().modify(|_, w| w.usr().set_bit());
+
+        #[cfg(not(esp32s2))]
```
Comment (MabezDev) [inline esp-hal/src/spi/slave.rs:442]: Could we get some explanation as to why all the chips apart from esp32s2 are started here, but esp32s2 is started earlier? This is quite hard to follow when looking at the driver for the first time.

File: esp-hal/src/spi/slave.rs [hunk 3]
```diff
@@ -464,20 +461,30 @@ where
...
+
         tx.prepare_transfer_without_start(self.dma_peripheral(), false, ptr, len)?;
+        #[cfg(esp32s2)]
+        tx.start_transfer()?;
 
-        self.clear_dma_interrupts();
         reset_dma_before_usr_cmd(reg_block);
 
-        // On the esp32, all full-duplex transfers are single, and all half-duplex
+        // Note: On the esp32, all full-duplex transfers are single, and all half-duplex
```
Comment (MabezDev) [inline esp-hal/src/spi/slave.rs:471]: I guess if we're not supporting the esp32, we don't need this comment and the other ones like it?

**Review by MabezDev — CHANGES_REQUESTED** [review]

**Comment by MabezDev** [comment]
Given that the example is _almost_ a test, could we add some test cases? I think we will need some hardware changes as we'll need more pins wired up.

**Comment by SergioGasquez** [comment]
> Given that the example is almost a test, could we add some test cases? I think we will need some hardware changes as we'll need more pins wired up.

This will be fun. See https://github.com/esp-rs/esp-hal/issues/1332#issuecomment-2059093479 to have some reference when choosing the new pins so we can avoid having too many `cfg`s arround

**Comment by MabezDev** [comment]
> This will be fun. See https://github.com/esp-rs/esp-hal/issues/1332#issuecomment-2059093479 to have some reference when choosing the new pins so we can avoid having too many cfgs arround

Maybe we can do this in another PR if it will be a pain, but it seems like something we can test in the future at least which would be nice!

File: esp-hal/src/spi/slave.rs (hunk 1 above)
Comment (JurajSadel) [reply esp-hal/src/spi/slave.rs:426]: It was moved called few lines later because of ESP32 - since we are not adding ESP32 I reverted this change.

File: esp-hal/src/spi/slave.rs (hunk 2 above)
Comment (JurajSadel) [reply esp-hal/src/spi/slave.rs]: It was due to ESP32, it required special order of calls - it's not needed for S2, so I removed it

File: esp-hal/src/spi/slave.rs (hunk 3 above)
Comment (JurajSadel) [reply esp-hal/src/spi/slave.rs]: I wanted to leave it there for future reference if someone breve enough will want to add ESP32 support, removed.

**Comment by JurajSadel** [comment]
> > This will be fun. See [#1332 (comment)](https://github.com/esp-rs/esp-hal/issues/1332#issuecomment-2059093479) to have some reference when choosing the new pins so we can avoid having too many cfgs arround
> 
> Maybe we can do this in another PR if it will be a pain, but it seems like something we can test in the future at least which would be nice!

Let's add this in another PR, Tomas is on leave this week and we need to buy cables as well, in a meantime I can try to figure out the pins.

**Review by MabezDev — APPROVED** [review]
Thanks!
